web_poet/rules.py

```diff
@@ -236,10 +295,32 @@ def search(self, **kwargs) -> List[ApplyRule]:
             print(rules[0].instead_of)  # GenericPO
 
         """
+        rule_ids = set()
+
+        # Ignore the types since we're using None value as a quick look-up.
+
+        matcher = self._item_matchers.get(kwargs.get("to_return"))  # type: ignore[arg-type]
+        if matcher:
+            rule_ids.update(matcher.patterns.keys())
+
+        matcher = self._overrides_matchers.get(kwargs.get("instead_of"))  # type: ignore[arg-type]
+        if matcher:
+            if rule_ids:
+                # If both params are used then narrow down the rules.
+                rule_ids.union(matcher.patterns.keys())
+            else:
+                rule_ids.update(matcher.patterns.keys())
+
+        rules = [self._rules[id_] for id_ in rule_ids]
+
+        if len(kwargs) == 1 and rules:
```

Comment:
It seems len(kwargs) can be 2, and we still might be able to return the result (kwargs.keys() is {to_return, instead_of}).

What do you think about using something like `kwargs.keys() <= {"to_return", "instead_of"}` ?

Reply:
Good call. I've refactored it to support this. More importantly, we can also now have quicker look-ups when `None` is passed when searching for either `instead_of` or `to_return`.